hera_qm/xrfi.py

```diff
@@ -1053,6 +1053,70 @@ def flag_apply(uvf, uv, keep_existing=True, force_pol=False, history='',
         return net_flags
 
 
+def simple_flag_waterfall(data,  Kt=8, Kf=8, sig_init=5.0, sig_adj=2.0, edge_cut=0, chan_thresh_frac=1.0):
+    '''XRFI-lite: performs median and mean filtering on a single waterfall, with 
+    watershed expansion of flags, and spectral and temporal thresholding.
+    
+    Parameters
+    ----------
+    data : 2D numpy array of floats or complex numbers
+        Waterfall to use to find and flag outliers.
+    Kt : int
+        Time half-width of kernel for med/meanfilt.
+    Kf : int
+        Frequency half-width of kernel for med/meanfilt.
+    nsig_init : float
+        The number of sigma in the metric above which to flag pixels. Default is 5.
+    nsig_adj : float
+        The number of sigma to flag above for points near flagged points. Default is 2.
+    edge_cut : integer
+        Number of channels at each band edge to flag automatically.
+    chan_thresh_frac : float
+        Fraction of times flagged (excluding completely flagged integrations) above which
+        to flag an entire channel. Default 1.0 means no additional flags.
+
+    Returns
+    -------
+    flags : 2D numpy array of booleans
+        Final waterfall of flags with the same shape as the data.
+        
+    '''
+    # Perform medfilt-based RFI excision with watershed growth of flags
+    medfilt_metric = detrend_medfilt(data, Kt=Kt, Kf=Kf)
+    medfilt_flags = (np.abs(medfilt_metric) > sig_init)
+    medfilt_flags |= _ws_flag_waterfall(medfilt_metric, medfilt_flags, nsig=sig_adj)
+    
+    # Perform meanfilt-based RFI excision with watershed growth of flags
+    meanfilt_metric = detrend_meanfilt(data, flags=medfilt_flags, Kt=Kt, Kf=Kf)
+    flags = medfilt_flags | (np.abs(meanfilt_metric) > sig_init) 
+    flags |= _ws_flag_waterfall(meanfilt_metric, flags, nsig=sig_adj)
+    
+    # Perform spectral thresholding 
+    combined_metric = np.where(medfilt_flags, medfilt_metric, meanfilt_metric)
+    spec = np.nanmedian(combined_metric, axis=0)
+    zspec = modzscore_1d(spec, detrend=False)
+    zspec[~np.isfinite(zspec)] = sig_init
+    flags[:, _ws_flag_waterfall(np.abs(zspec), np.abs(zspec) >= sig_init, nsig=sig_adj)] = True
+
+    # Perform temporal thresholding
```

Comment:
What sort of temporal structure do we expect to show up in the averaged waterfall? This concerns me a little, because if it's leaking in from the temporal instability of a particular antenna, even if we flag the integrations that contain the discontinuity, there may be residual bandpass weirdness in integrations before/after the actual discontinuity. This isn't necessarily a criticism of flagging temporal structure, but more a commentary that if we have to flag temporal structure it means we are getting more of it in the averaged waterfall than we would hope, since dividing out a waterfall with that type of structure might explain why we were seeing so much temporal structure in your waterfalls. I'm not sure I have a specific fix in mind, mostly just wondering if you've already thought about this, or if I'm interpreting things wrong.

Reply:
This is a good question. If all antennas have a temporal discontinuity at the same time in the same way, this algorithm will attribute that to RFI rather than an antenna problem. When an integration is flagged, it's ignored in the collapsed spectra/metrics (and in a diff-based metric, np.diff will propagate to the nan to both diffs involving that pixel). I think that's consistent with what we're seeing, which is apparently uncorrelated temporal discontinuities between antennas. Also, this flagging algorithm is applied only after throwing out the most egregious antennas in the "median round," so the remainder shouldn't add much to the rumble of the antenna-averaged waterfall that we then flag on.

A separate question is whether temporal structure in one antenna, when we average all antennas together to produce a per-polarization waterfall to normalize out for the "Temporal Variability" and "Temporal Discontinuity" metrics, affects the measured temporal structure in the others. We expect that it should, at some level, but that's why I have these lines:
https://github.com/HERA-Team/hera_qm/blob/2192099f6964a52d2a86fa18dc33c7a8f788aacb/hera_qm/auto_metrics.py#L388-L392

After each round of flagging, I recompute the spectra that depend on waterfall normalization, removing them from the average, before re-running the Z-score calculation. This should help reduce the cross-contamination of bad antennas.

Reply:
This pretty much answers my question, and I think the way you've implemented things minimizes "structure leakage" well (and I can't think of a straightforward change that would make it better). One of the reasons for my concern is that when looking at the SSINS of this data we were seeing a lot of temporal structure, and SSINS does a mean over all baselines after the time-differencing steps, and I think even though the temporal structure isn't aligned between antennas it was still bright enough to show up in the mean even when it was only present in one antenna. It should affect all antennas roughly the same, so I don't think this will really affect which antennas we flag, but I do think it could produce a bias where the worst antennas look even worse because they are being compared against an average waterfall with structure. I'm fine with moving forward with things as they are for now though, and just keeping an eye on whether this seems to be causing problems.